Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkHierarchyRootComponent.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/Component/Component.h>
+#include <AzCore/Component/TransformBus.h>
+#include <Source/AutoGen/NetworkHierarchyRootComponent.AutoComponent.h>
+#include <Multiplayer/Components/NetworkHierarchyBus.h>
+
+namespace Multiplayer
+{
+    //! @class NetworkHierarchyRootComponent
+    //! @brief Component that declares the top level entity of a network hierarchy.
+    /*
+     * Call @GetHierarchyChildren to get the list of hierarchical entities.
+     * A network hierarchy is meant to be a small group of entities. You can control the maximum supported size of
+     * a network hierarchy by modifying CVar @bg_hierarchyEntityMaxLimit.
+     *
+     * A root component marks either a top most root of a hierarchy, or an inner root of an attach hierarchy.
+     */
+    class NetworkHierarchyRootComponent final
+        : public NetworkHierarchyRootComponentBase
+        , public NetworkHierarchyRequestBus::Handler
+        , protected AZ::TransformNotificationBus::MultiHandler
+    {
+    public:
+        AZ_MULTIPLAYER_COMPONENT(Multiplayer::NetworkHierarchyRootComponent, s_networkHierarchyRootComponentConcreteUuid, Multiplayer::NetworkHierarchyRootComponentBase);
+
+        static void Reflect(AZ::ReflectContext* context);
+        static void GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required);
+        static void GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided);
+        static void GetIncompatibleServices(AZ::ComponentDescriptor::DependencyArrayType& incompatible);
+
+        //! NetworkHierarchyRootComponentBase overrides.
+        //! @{
+        void OnInit() override;
+        void OnActivate(Multiplayer::EntityIsMigrating entityIsMigrating) override;
+        void OnDeactivate(Multiplayer::EntityIsMigrating entityIsMigrating) override;
+        //! @}
+
+        //! NetworkHierarchyRequestBus overrides.
+        //! @{
+        bool IsHierarchicalRoot() const override;
+        bool IsHierarchicalChild() const override;
+        AZStd::vector<AZ::Entity*> GetHierarchicalEntities() const override;
+        AZ::Entity* GetHierarchicalRoot() const override;
+        //! @}
+
+    protected:
+        //! AZ::TransformNotificationBus::Handler overrides.
+        //! @{
+        void OnParentChanged(AZ::EntityId oldParent, AZ::EntityId newParent) override;
+        void OnChildAdded(AZ::EntityId child) override;
+        void OnChildRemoved(AZ::EntityId childRemovedId) override;
+        //! @}
+
+        void SetTopLevelHierarchyRootEntity(AZ::Entity* hierarchyRoot);
```

Comment:
Might be worth a small function comment explaining how you can have something higher than a root? since in most cases "top level" and "root" are synonymous but not in our case when we can have nested/higher roots.

Reply:
It's a protected method, so users don't have access to it, nor can they can it themselves. I think I would rather add a note to hierarchies of hierarchies to the class header instead. Good call.

Reply:
Actually class header already states the following:

> A root component marks either a top most root of a hierarchy, or an inner root of an attach hierarchy.

Reply:
This is a protected method that users dont have access to. And the clear header already states:

> A root component marks either a top most root of a hierarchy, or an inner root of an attach hierarchy.